streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1038,6 +1053,9 @@ private void buildProcessorNode(final Map<String, ProcessorNode> processorMap,
                                 createChangelogTopicConfig(stateFactory, topicName);
                             stateChangelogTopics.put(topicName, internalTopicConfig);
                         }
+                        hasStateToRestore = true;
```

Comment:
I thought about an alternative idea here and just leave it to your call whether you like it or not: I think we can modify line 1047 above to only `if (stateFactory.users().contains(node))`, since the `storeToChangelogTopic` should contain the source-changelog topics already, so we can just add it to the `stateChangelogTopics` of the TopicsInfo.

Then, in StreamsPartitionAssignor, before `prepareTopic(changelogTopicMetadata);` we only prepare the topics as `changelog-topics minus source-topics`. In this way we do not need this flag either.

My rationale is that this is more consistent with the name `stateChangelogTopics` (a source-changelog topic should be included in this map, as well as in `source-topics`, we just need to let the partition-assignor to be aware of the possible overlapping).

Reply:
Nice idea, that lets us simplify the special handling for optimized source KTables

Reply:
Somehow one of my comment was lost... I also comment that `hasStateWithChangelogs` can then be implemented as `! stateChangelogTopics.empty()`.

Reply:
No worries, I had the same idea and simplified the whole thing. Decided to add a public method to `TopicsInfo` to sift out any optimized source tables rather than add that logic directly to the `StreamsPartitionAssignor` (which is complicated enough as is)